core/src/main/java/org/apache/iceberg/rest/OAuth2Util.java

```diff
@@ -0,0 +1,142 @@
+  private static final String CATALOG = "catalog";
+
+  // Client credentials flow
+  private static final String CLIENT_ID = "client_id";
+  private static final String CLIENT_SECRET = "client_secret";
+
+  // Token exchange flow
+  private static final String SUBJECT_TOKEN = "subject_token";
+  private static final String SUBJECT_TOKEN_TYPE = "subject_token_type";
+  private static final String ACTOR_TOKEN = "actor_token";
+  private static final String ACTOR_TOKEN_TYPE = "actor_token_type";
+  private static final Set<String> VALID_TOKEN_TYPES = Sets.newHashSet(
+      "urn:ietf:params:oauth:token-type:access_token",
+      "urn:ietf:params:oauth:token-type:refresh_token",
+      "urn:ietf:params:oauth:token-type:id_token",
+      "urn:ietf:params:oauth:token-type:saml1",
+      "urn:ietf:params:oauth:token-type:saml2",
+      "urn:ietf:params:oauth:token-type:jwt");
+
+  public static boolean isValidScopeToken(String scopeToken) {
+    return VALID_SCOPE_TOKEN.matcher(scopeToken).matches();
+  }
+
+  public static List<String> parseScope(String scope) {
+    return SCOPE_DELIMITER.splitToList(scope);
+  }
+
+  public static String toScope(List<String> scopes) {
+    return SCOPE_JOINER.join(scopes);
+  }
+
+  public static Map<String, String> tokenExchangeRequest(String subjectToken, String subjectTokenType,
+                                                         List<String> scopes) {
+    return tokenExchangeRequest(subjectToken, subjectTokenType, null, null, scopes);
+  }
+
+  public static Map<String, String> tokenExchangeRequest(String subjectToken, String subjectTokenType,
+                                                         String actorToken, String actorTokenType,
+                                                         List<String> scopes) {
+    Preconditions.checkArgument(VALID_TOKEN_TYPES.contains(subjectTokenType),
+        "Invalid token type: %s", subjectTokenType);
+    Preconditions.checkArgument(actorToken == null || VALID_TOKEN_TYPES.contains(actorTokenType),
+        "Invalid token type: %s", actorTokenType);
+
+    ImmutableMap.Builder<String, String> formData = ImmutableMap.builder();
+    formData.put(GRANT_TYPE, TOKEN_EXCHANGE);
+    formData.put(SCOPE, toScope(scopes));
+    formData.put(SUBJECT_TOKEN, subjectToken);
+    formData.put(SUBJECT_TOKEN_TYPE, subjectTokenType);
+    if (actorToken != null) {
+      formData.put(ACTOR_TOKEN, actorToken);
+      formData.put(ACTOR_TOKEN_TYPE, actorTokenType);
+    }
+
+    return formData.build();
+  }
+
+  private static Pair<String, String> parseCredential(String credential) {
+    Preconditions.checkNotNull(credential, "Invalid credential: null");
+    List<String> parts = CREDENTIAL_SPLITTER.splitToList(credential);
+    switch (parts.size()) {
+      case 2:
+        // client ID and client secret
+        return Pair.of(parts.get(0), parts.get(1));
+      case 1:
+        // client ID
+        return Pair.of(null, parts.get(0));
```

Comment:
Is it intended that `client ID` is sometimes pair[0] and sometimes pair[1]?

That's how it looks presently based on the comments and return statements.

Reply:
No. The secret is always present but the client ID may not be.

Reply:
Oh. That's what I had thought. Then I believe the comment on line 113 should say `// client secret` and not `// client ID`, if that's what's present (and get's placed into the second part of the pair every time).

Reply:
Yes, it should. Fixed in #4833.